### PR TITLE
feat: Require login

### PR DESCRIPTION
**Pull request by KevinJBoyer**
## Ticket

https://navalabs.atlassian.net/browse/DST-234

## Changes
 - Add `shared.py` so code isn't reconstructing expensive instances like AppConfig or SentenceTransformer unnecessarily. Maybe a better name would be `shared.py`?
 - Add `login.py` that adds a login page when `GLOBAL_PASSWORD` and `CHAINLIT_AUTH_SECRET ` are set

## Context for reviewers
 - This automatically integrates with LiteralAI for tracking by the email field -- see screenshot below

## Testing
For local testing, set `GLOBAL_PASSWORD` and `CHAINLIT_AUTH_SECRET` in your .env and restart the app (`make start`). You may want to use incognito windows to avoid clearing your cookies each time.

Screenshot showing integration with LiteralAI (when this API key is also set):
Screenshot 2024-07-02 at 11 12 05 AM

### CODE REVIEW

File: app/src/app_config.py [hunk 1]
```diff
@@ -3,3 +3,4 @@
 
 class AppConfig(PydanticBaseEnvConfig):
     embedding_model: str = "multi-qa-mpnet-base-dot-v1"
+    global_password: str | None = None
```
Comment (KevinJBoyer) [inline app/src/app_config.py:6]: PydanticBaseEnvConfig  ensures these defaults are overridden by any environment variables that are set

File: infra/app/app-config/env-config/environment-variables.tf [hunk 2]
```diff
@@ -19,11 +19,15 @@ locals {
...
   # }
   secrets = {
-    # Example generated secret
-    # RANDOM_SECRET = {
-    #   manage_method     = "generated"
-    #   secret_store_name = "/${var.app_name}-${var.environment}/random-secret"
-    # }
+    CHAINLIT_AUTH_SECRET = {
+      manage_method     = "generated"
+      secret_store_name = "/${var.app_name}-${var.environment}/CHAINLIT_AUTH_SECRET"
```
Comment (yoomlam) [inline infra/app/app-config/env-config/environment-variables.tf:24]: Where is `CHAINLIT_AUTH_SECRET` used?

**Review by yoomlam — CHANGES_REQUESTED** [review]

File: app/src/cache.py
[the remark this answers is not in the document]
Comment (KevinJBoyer) [reply app/src/cache.py:1]: Will make this change!

File: infra/app/app-config/env-config/environment-variables.tf (hunk 2 above)
Comment (KevinJBoyer) [reply infra/app/app-config/env-config/environment-variables.tf:24]: Added a comment to the file explaining this (it's used by Chainlit to sign the authentication tokens)

File: app/tests/src/test_login.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+import chainlit.config
+import src.cache
+from src.login import login_callback, require_login
+
+
+def test_require_login_no_password(monkeypatch):
+    if "GLOBAL_PASSWORD" in os.environ:
+        monkeypatch.delenv("GLOBAL_PASSWORD")
+    src.cache._app_config = None
```
[the remark this answers is not in the document]
Comment (KevinJBoyer) [reply app/tests/src/test_login.py]: Added a comment to the first instance of this in the file!

File: app/src/cache.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from sentence_transformers import SentenceTransformer
+
+from src.app_config import AppConfig
+
+_app_config: AppConfig | None = None
+
+
+def get_appconfig() -> AppConfig:
+    global _app_config
```
[the remark this answers is not in the document]
Comment (KevinJBoyer) [reply app/src/cache.py]: TIL about the built in cache decorators-- updated the code to use this!

File: app/tests/src/test_login.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    require_login()
```
[the remark this answers is not in the document]
Comment (KevinJBoyer) [reply app/tests/src/test_login.py:15]: Yeah, ideally I'd like to have some kind of integration test of `chainlit.py`. Maybe a future improvement?

**Review by yoomlam — APPROVED** [review]

**Pull request merged**